lib/auth/middleware.go

```diff
@@ -210,16 +301,93 @@ type AuthMiddleware struct {
 	// if set, will accept certificates with non-limited usage,
 	// and usage exactly matching the specified values.
 	AcceptedUsage []string
+	// Limiter is a rate and connection limiter
+	Limiter *limiter.Limiter
 }
 
 // Wrap sets next handler in chain
 func (a *AuthMiddleware) Wrap(h http.Handler) {
 	a.Handler = h
 }
 
+// UnaryInterceptor is GPRC unary interceptor that authenticates requests
+// and passes the user information as context metadata
+func (a *AuthMiddleware) UnaryInterceptor(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
+	peerInfo, ok := peer.FromContext(ctx)
+	if !ok {
+		return nil, trail.ToGRPC(trace.AccessDenied("missing authentication"))
+	}
+	// Limit requests per second and simultaneous connection by client IP.
+	clientIP, _, err := net.SplitHostPort(peerInfo.Addr.String())
```

Comment:
What is peer address for an IoT node?  Doesn't the connection technically come from the proxy in that case?

Reply:
You are right - the address will be proxy in case of IOT nodes. This problem is not limited to grpc, happens in HTTPs rate / conn limiter as well. I will think how to address it if it shows up in production at some point.

Reply:
I created https://github.com/gravitational/teleport/issues/4374 to track this separately